promote "initial_condition" parameter to the variable action level

rwcarlsen opened the pull request:
This helps with external-tool validation of input file syntax.  And
makes things more clean anyway - the parameter is only used by the
action - it should be owned by the action.  Fixes #14231.

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

rwcarlsen committed:
promote "initial_condition" parameter to the variable action level

This helps with external-tool validation of input file syntax.  And
makes things more clean anyway - the parameter is only used by the
action - it should be owned by the action.  Fixes #14231.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/623133/) on f220720 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16178/site)

This comment will be updated on new commits.

rwcarlsen committed:
Describe when actions provide/set parameters for their (sub)objects

This allows input file syntax validation to account for some required
parameters not being provided in objects' input blocks.

Fixes #14232

rwcarlsen committed:
tweak special param name. Also fix same issue in output action.

fixes #15350

rwcarlsen commented:
@brandonlangley - does this look good to you - does it seem to address the issues in a way that makes sense?

brandonlangley commented:
@rwcarlsen -

Thanks a lot for this PR.

These changes look good and it looks like they will address all of https://github.com/idaholab/moose/issues/14231, https://github.com/idaholab/moose/issues/14232, https://github.com/idaholab/moose/issues/15350.

This afternoon I'm going to build this out to:

 - make sure `initial_condition` now shows up in the JSON for `Variables` and `AuxVariables`

 - check if there are any other contexts that need the `_object_params_set_by_action` besides
   - `Variables/InitialCondition`-->`variable`
   - `Outputs `-->`file_base`

Then, I'll approve. Sorry about the delay in reviewing with Thanksgiving vacation last week.

dschwen commented:
As a side note, the `initial_condition` parameter did not show up in the auto complete plugin either, until this recent change: https://github.com/dschwen/autocomplete-moose/commit/2545c781e5a501bd18aa7e658e7a8b70379acea2 so I'd argue the moose syntax JSON was well formed all along (nonetheless moving said parameter seems like the right thing to do).
